<a href="https://colab.research.google.com/github/michalis0/Business-Intelligence-and-Analytics/blob/master/10%20-%20Classification2/Exercises/Exercises_Classification_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Classification 2

## KNN and Decision Trees

In [1]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import collections  as mc
%load_ext autoreload
%autoreload 2
import pandas as pd 
import seaborn as sns
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
sns.set_style("white")

In [2]:
np.random.seed = 72

### Load data

We will use the same `heart disease` dataset from last week. Here's a description of the attributes in the data-set:

> 1. age
> 2. sex
> 3. chest pain type (4 values)
> 4. resting blood pressure
> 5. serum cholestoral in mg/dl
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved
> 9. exercise induced angina
> 10. target: 1: presence 0: absence of a heart disease

In [4]:
# data-set: heart.csv
df = pd.read_csv("https://media.githubusercontent.com/media/michalis0/Business-Intelligence-and-Analytics/master/data/heart.csv")
df.head()

age sex cp  trestbps  chol  fbs restecg  thalach exang  target
0   63   M  D       145   233  yes       A      150    no       1
1   37   M  C       130   250   no       B      187    no       1
2   41   F  B       130   204   no       A      172    no       1
3   56   M  B       120   236   no       B      178    no       1
4   57   F  A       120   354   no       B      163   yes       1

In [5]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn import datasets

### K-neirest neighbor classification

Let's start with only 2 features: age and maximum heart rate achieved (thalach). Define your features and target variable. 

In [ ]:
X = "YOUR CODE HERE"
y = "YOUR CODE HERE"

Split your data set into train and test subsets. 

In [ ]:
from sklearn.model_selection import train_test_split
"YOUR CODE HERE" = "YOUR CODE HERE"(X, y, test_size=0.2, random_state=72)

Define your model. Try to use a KNN ( with n=3)

In [ ]:
#Create model
model = "YOUR CODE HERE"

Fit your model now using the train set. 

In [ ]:
model.fit("YOUR CODE HERE")

Compare your train and test accuracy for your model.

In [ ]:
# train accuracy with CV
"YOUR CODE HERE"

In [ ]:
# test accuracy with CV
"YOUR CODE HERE"

Calculate the base rate

In [ ]:
#Compute the base rate
#Base rate = (most frequent class)/(total observations)
nbr_heart_disease = df.loc[df["YOUR CODE HERE"] == 1].shape[0]
print("#Heart disease = ", "YOUR CODE HERE")

nbr_no_heart_disease = df.loc[df["YOUR CODE HERE"] == 0].shape[0]
print("#No heart disease = ", "YOUR CODE HERE")

print("Baserate = ", "YOUR CODE HERE")

Show the confusion matrix

In [ ]:
#Confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix("YOUR CODE HERE")

For which K (number of neighbors), p-value and weight do we obtain the best accurary?

In [ ]:
n_neighbors = [2,3,4,5,6,7]
p_values = [1,2,3]
weights = ['uniform', 'distance']
for n in n_neighbors:
    for p in p_values:
        for w in weights:
            model = KNeighborsClassifier(n_neighbors=n, p=p, weights=w)
            model.fit("YOUR CODE HERE")
            print("accuracy for k = " + str(n) + ", p = " + str(p) + ", weight = " + w + " : " + str(model.score("YOUR CODE HERE")))

Now, let's try to predict the class and probability of correct classification for Age = 50, Thalach = 130 and Age = 70, Thalach = 160

In [ ]:
print(model.predict("YOUR CODE HERE"))
print(model.predict_proba("YOUR CODE HERE"))

Now let's try more numerical features and see if the accuracy improuves.
We will use now "age", "thalach", "trestbps" and "chol". 
Define your features and your target variable. 

In [ ]:
X = "YOUR CODE HERE"
y = "YOUR CODE HERE"

Split your data set into train and test subsets.

In [ ]:
"YOUR CODE HERE" = "YOUR CODE HERE"(X, y, test_size=0.2, random_state=72)

Standardize your data. 

In [ ]:
standardize = StandardScaler()
"YOUR CODE HERE"

In [ ]:
X_train = "YOUR CODE HERE"
X_test = "YOUR CODE HERE"

Fit your model using the train data. Let's use the KNN here. 

In [ ]:
model.fit("YOUR CODE HERE")

Compare your train and test accurary. 

In [ ]:
# train accuracy
"YOUR CODE HERE"

In [ ]:
# test accuracy
"YOUR CODE HERE"

In [ ]:
#Confusion matrix
from sklearn.metrics import confusion_matrix
"YOUR CODE HERE"

### Decision tree classification

We will now use the decision tree algorithm for classification.

Import the necessary libraries for decision tree classification, define the model and train your model. The test it. What is the accuracy of your model (using the test set)?

In [ ]:
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
model = DecisionTreeClassifier(max_depth = 1)
model.fit("YOUR CODE HERE")
print(model.score("YOUR CODE HERE"))

How can you improve your model? Find the lowest depth with for an accuracy of 1. 

In [ ]:
#function to find lowest depth (less complicated model) keeping an accuracy of 1
depth = 5
model = DecisionTreeClassifier(max_depth = depth)
model.fit("YOUR CODE HERE")
accuracy = model.score("YOUR CODE HERE")
while accuracy == 1:
    depth -= 1
    model = DecisionTreeClassifier(max_depth = depth)
    model.fit("YOUR CODE HERE")
    accuracy = model.score(X_test, y_test)
print(depth+1)

Plot the decision tree and the decision boundaries for a model using 'age' and 'thalach' as features. How do you interpret these visualisations?

In [ ]:
import numpy as np
import matplotlib.pyplot as plt

from sklearn.tree import DecisionTreeClassifier, plot_tree

plot_colors = "br"
plot_step = 0.02


X = X[["YOUR CODE HERE"]].values

clf = Decision...().fit(X, y)

x_min, x_max = X[:, 0].min() - 1, X[:, 0].max() + 1
y_min, y_max = X[:, 1].min() - 1, X[:, 1].max() + 1
xx, yy = np.meshgrid(np.arange(x_min, x_max, plot_step), np.arange(y_min, y_max, plot_step))
plt.tight_layout(h_pad=0.5, w_pad=0.5, pad=2.5)

Z = clf.predict(np.c_[xx.ravel(), yy.ravel()])
Z = Z.reshape(xx.shape)
cs = plt.contourf(xx, yy, Z, cmap=plt.cm.RdYlBu)

plt.xlabel(...)
plt.ylabel(...)

# Plot the training points
for i, color in zip(range(2), plot_colors):
    idx = np.where(y == i+1)
    plt.scatter(X[idx, 0], X[idx, 1], c=color,
                cmap=plt.cm.RdYlBu, edgecolor='black', s=15)

plt.suptitle("Decision surface of a decision tree using paired features")
plt.axis("tight")

plt.figure(figsize=(11,7))
clf = "YOUR CODE HERE"().fit(X, y)
plot_tree(clf, filled=True)
plt.show()